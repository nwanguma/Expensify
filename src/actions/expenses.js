import uuid from 'uuid';
import database from '../firebase/firebase';

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
});

export const startAddExpense = (expenseData = {}) => {
  //this dispatch argument is provided by thunk
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;
    const expense = { description, note, amount, createdAt };

    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }))
    })
  }
}

// REMOVE_EXPENSE
export const removeExpense = ({ id }) => ({
  type: 'REMOVE_EXPENSE',
  id
});


// REMOVE_EXPENSE_ASYNC
export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({ id }));
    })
  }
}

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});

// EDIT_EXPENSE_ASYNC
export const startEditExpense = (id, updates) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id, updates));
    })
  }
}

// SET_EXPENSE
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
})

// SET_EXPENSES_ASYNC
export const startSetExpenses = () => {
  return (dispatch) => {
    return database.ref('expenses').once('value').then((snapshot) => {
      const expenses = [];

      snapshot.forEach(childSnapshot => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });
      dispatch(setExpenses(expenses));
    });
  };
};
