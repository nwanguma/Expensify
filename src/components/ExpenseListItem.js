import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';
import { connect } from 'react-redux';
import { startRemoveExpense } from '../actions/expenses';

class ExpenseListItem extends React.Component {
  constructor(props) {
    super(props);
  }
  onRemove = () => {
    this.props.startRemoveExpense({ id: this.props.id });
  }
  render() {
    return (
      <div className="expense-item-container">
        <Link to={`/edit/${this.props.id}`} className="expense__content 
        expense__content--link" >
          <h3 className="expense__text expense__content--description">
            {this.props.description}</h3>
        </Link>
        <span className="expense__text expense__content--amount">
          {`â‚¦${numeral(this.props.amount / 100).format('0,0.00')}`}
        </span>
        <span className="expense__text expense__content--date">
          {moment(this.props.createdAt).format('MMMM Do, YYYY')}
        </span>
        <a onClick={this.onRemove} className="btn btn--remove" role="button">x</a>
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => ({
  startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
});

export default connect(undefined, mapDispatchToProps)(ExpenseListItem);
