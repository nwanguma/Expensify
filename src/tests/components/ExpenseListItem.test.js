import React from 'react';
import { shallow } from 'enzyme';
import ExpenseListItem from '../../components/ExpenseListItem';
import expenses from '../fixtures/expense';

test('Should render the expenseListItem component with props', () => {
  const wrapper = shallow(<ExpenseListItem {...expenses[0]} />);
  expect(wrapper).toMatchSnapshot();
})