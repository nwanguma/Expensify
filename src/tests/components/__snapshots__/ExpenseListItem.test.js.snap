// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render the expenseListItem component with props 1`] = `
<div>
  <Link
    to="/edit/ioeyuienjdsjaoiihepo"
  >
    <h3>
      Drinks
    </h3>
  </Link>
  <p>
    $67.00
    -
    February 1st, 1970
  </p>
</div>
`;
